fix: NetworkBehaviour spam generated when LogLevel is set to developer

NoelStephensUnity opened the pull request:
When the LogLevel is set to developer, log warnings were spamming even when the NetworkBehaviour.NetworkObject was not null.  This PR addresses the issue such that the log warnings are still generated when there is no NetworkObject assigned to a NetworkBehaviour and are not generated when the NetworkBehaviour.NetworkObject is not null. 

[MTT-2289](https://jira.unity3d.com/browse/MTT-2289)

### PR Checklist
- [x] Have you added a backport label (if needed)? For example, the `type:backport-release-*` label. After you backport the PR, the label changes to `stat:backported-release-*`.
- [X] Have you updated the changelog? Each package has a `CHANGELOG.md` file.

## Changelog
### com.unity.netcode.gameobjects
-Fixed: When the LogLevel is set to developer NetworkBehaviour generates warning messages when it should not

## Testing and Documentation
* Includes integration tests.

NoelStephensUnity committed:
fix

fixing issue with spam when setting NetworkManager.LogLevel = LogLevel.Developer.

NoelStephensUnity committed:
test

This test verifies the fix.  When the LogLevel is set to developer, log warnings should be generated when there is no NetworkObject assigned to a NetworkBehaviour and should not be generated when there is a valid NetworkObject.

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
Merge branch 'develop' into fix/NetworkBehaviour-Spam-DeveloperLogging

TwoTenPvP reviewed (approved):
Other than the hardcoded log, which is fine since this wil fail instead of pass if changed. This looks good!

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs:
```diff
@@ -304,8 +304,7 @@ public NetworkObject NetworkObject
 
-                if (m_NetworkObject == null || NetworkManager.Singleton == null ||
-                    (NetworkManager.Singleton != null && !NetworkManager.Singleton.ShutdownInProgress))
+                if (m_NetworkObject == null && (NetworkManager.Singleton == null || !NetworkManager.Singleton.ShutdownInProgress))
```
nit: While you're at it it may be nice to explain why we care about shutdown here

NoelStephensUnity replied:
That was a good idea and while I was in there I realized that the original LogLevel check was <= as opposed to just <.  I believe it got set because it was spamming when the default Normal LogLevel was set.

JesseOlmer commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkBehaviourGenericTests.cs:
```diff
@@ -0,0 +1,60 @@
+        /// particular test if new tests are added to this class that do require clients
+        /// </summary>
+        [Test]
+        public void ValidateNoSpam()
```
Do you really need the overhead of the multiinstancehelpers and actually starting a host in this test? Feels like you should be able to do this closer to a "unit" test... something like this:

```
        [Test]
        public void NetworkObjectGetterLogsMissingNetworkObject()
        {
            // Arrange NB with no NO, and Developer logging
            var gameObject = new GameObject();
            var nb = gameObject.AddComponent<EmptyNetworkBehaviour>();
            var nm = gameObject.AddComponent<NetworkManager>();
            nm.SetSingleton();
            nm.LogLevel = LogLevel.Developer;

            // Act
            var _ = nb.NetworkObject;

            // Assert - expected warning fired, tested generically to make the test less cumbersome
            LogAssert.Expect(LogType.Warning, new Regex(@"\[Netcode\].+"));
            LogAssert.NoUnexpectedReceived();
        }

        [Test]
        public void NetworkObjectGetterDoesNotLog()
        {
            // Arrange NB with an NO, and Developer logging
            var gameObject = new GameObject();
            var nb = gameObject.AddComponent<EmptyNetworkBehaviour>();
            gameObject.AddComponent<NetworkObject>();
            var nm = gameObject.AddComponent<NetworkManager>();
            nm.SetSingleton();
            nm.LogLevel = LogLevel.Developer;

            // Act
            var _ = nb.NetworkObject;
            
            LogAssert.NoUnexpectedReceived();
        }
```

or does that just cause problems "cleaning up" after the test?

NoelStephensUnity replied:
Yeah, I was thinking about making it a simple Unit test, but then realized we have no "generic" test for NetworkBehaviour where team members can add additional "generic tests and fix verifications".
I am more than happy to migrate this over to just a unit test (let me know your preference), but I would basically need to replicating the a bunch of the same code in the Setup and TearDown methods that comes with BaseMultiInstanceTest.

JesseOlmer replied:
It's up to you - I already approved the PR 👍

JesseOlmer reviewed (approved):
lgtm

NoelStephensUnity committed:
style and update

Added notes about the ShutdownInProgress check as well reverted the NetworkLog.CurrentLogLevel requirement to generate this message back to its original <= value so this warning will be logged by default if the proper conditions to generate the message exist.
